src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs

```diff
@@ -59,6 +60,22 @@ internal static class ProtocolConversions
             { WellKnownTags.NuGet, LSP.CompletionItemKind.Text }
         };
 
+        public static Uri GetUriFromFilePath(string filePath)
+        {
+            if (filePath is null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            // Remove preceding slash if we're on Window as it's an invalid URI.
+            if (filePath.StartsWith("/") && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                filePath = filePath.Substring(1);
+            }
+
+            return new Uri(filePath, UriKind.Absolute);
+        }
```

Comment:
i hate this entire function.

Reply:
wiling to take suggestions on how to improve this - it's here because of how razor file paths are normalized to have the same representation on mac/linux/windows.
@NTaylorMullen can clarify more.

Reply:
First off it's definitely not ideal but there's a need to handle "bad" line pragmas anyhow.

The input into this is from a line pragma and Razor happens generates line pragmas with a leading `/` to keep track of files in a platform independent way. A lot of the file APIs that the runtime uses allow a prefixed `/` and so we prefix all of our paths with it to enable having the same path representation on Windows Mac and Linux.
 
Thinking from your perspective, you all will need to have a flow where you're unable to convert a line pragma into a Uri anyhow. Would it be awful to maintain this attempt small normalization? I ask because it'd be a massive change for our language server to move away from normalizing all our platform specific file paths to look the same.

Reply:
i don't hate the URI.  But i can't stand that we're having to munge the paths ourselves, or have to know anything about what platform we're on or the rules about paths.  

> Would it be awful to maintain this attempt small normalization? 

Basically, i have no idea how to review this function.  Is it correct?  I have no idea.  Is it complete?  I have no idea.  Will it throw and bring down something important?  I have no idea.  

> A lot of the file APIs that the runtime uses allow a prefixed / and so we prefix all of our paths with it to enable having the same path representation on Windows Mac and Linux.

Why can't URIs handle this?  How are URIs expected to work on Mac and Linux then?

Reply:
> Why can't URIs handle this?

That would be the ideal solution; however, changing URIs is a much more difficult change to do than to ensure we normalize the input to a `Uri` construction. 

As for how URIs are expected to work on Mac/Linux: I'm not fully confident in the internals of Uri but filepaths being prefixed with `/` is the normal in Linux/Mac.

> Basically, i have no idea how to review this function. Is it correct? I have no idea. Is it complete? I have no idea. Will it throw and bring down something important? I have no idea.

I imagine you have a centralized Uri normalization flow when attempting to go from file path => Uri given users can put pretty much anything into a line pragma. 

This function's "correctness" and "completness" should be evaluated on its return value; aka, if it looks like a valid Uri then the Roslyn system shouldn't fall over even if the addressable location of the Uri cannot be found (that's the worse case).  

"Will it bring down something important" i'm making the statement of "if it looks like a valid Uri then Roslyn shouldn't fall over" because Uri's are sent to your language server from the LSP client and are outside of Roslyn's control in terms of what format they're sent. Aka, you already need to have resilience to the "wellformedness" of the Uri.

Reply:
Yeah, this bit terrifies me as well, simply because it means this bit has to be special. Special is bad. Special leads to bugs. How do we keep a proper test suite to ensure that we don't screw this up, and this function works right?

The OS checking also worries me,simply because our LSP shouldn't depend on OS. If we had a unit test on this (which we should have no matter what....) you couldn't get this to pass on both sides without conditioning the unit test, right? And if LSP is being used cross-machine, what should be happening then?

Reply:
I've filed an [issue](https://github.com/dotnet/aspnetcore/issues/19948) on our side to stop normalizing windows paths down to `/C:/path/to/something.razor`. Until we're able fix this issue (probably by end of April) would it be possible to take on the burden of this normalization?

Reply:
I've very hestiant about that, unless we have confirmation that we can rip this code out at a certain date no matter what.  The main problem is that we have tons of cases of teams asking us to do this sort of thing, but then never following through on their end. (why would they?  the problem is now 'gone' because of the code in Roslyn).  So we have hacks and hacks that have to live forever.

My strong preference is that if this is going to be fixed on the razor side that we just fix that there.  Then we avoid the problem entirely.  What's the downside of just waiting on htat?  End of april is only 6 weeks away.  Is it necessary to have things fixed prior to that (and, if so, couldn't we just fix it faster in razor)?

Reply:
> The main problem is that we have tons of cases of teams asking us to do this sort of thing, but then never following through on their end.

I totally understand! Some differences in our case though is that I personally have the decision power on what we do for which release and in the coming months we're going to be interacting a LOT. I personally will ensure this hack doesn't live forever.



> My strong preference is that if this is going to be fixed on the razor side that we just fix that there. Then we avoid the problem entirely. What's the downside of just waiting on htat? End of april is only 6 weeks away.

So sadly begin/mid April is when we're code completing for \\build in the attempt to deliver a quality "cloud" Razor experience. This item (delivering LSP diagnostics for Razor files) is one of our deliverables for that timeframe.

> couldn't we just fix it faster in razor?

Currently we're slammed with items attempting to get this \\Build scenario to light up and mucking with how our LanguageServer understands file paths in Razor is a significantly large item. We will absolutely do it but given its size we wont have time to do it prior to locking down for \\build.


--------------

So all that being said

> confirmation that we can rip this code out at a certain date no matter what.

To ensure we have enough time to complete the work and to account for "massiveness" on our end to fix it; can we compromise and say you all can 100% rip this out at the end of May (we will push for end of April though) and break the Razor experience? Oh and IF this issue causes any other issues you can rip it out immediately and we can wait for the Razor fix (losing Razor diagnostics of course, but don't want to cause a burden).

I've assigned the work item to our current and next iterations so it wont be lost and hell, even this convo makes it stick top-of-mind that we need to fix this 😄 .

If we can agree on an end-of-may no matter what rip-out date I'll update the issue on our end to reflect the agreement.

Reply:
> To ensure we have enough time to complete the work and to account for "massiveness" on our end to fix it; can we compromise and say you all can 100% rip this out at the end of May (we will push for end of April though) and break the Razor experience?

Unfortunately "you can just break us" still turns into a "no we can't" once the reality of ship schedules and perhaps automated tests comes in. But the idea at least I can get behind.

Thanks for filing a tracking item.

Reply:
> Currently we're slammed with items attempting to get this \Build scenario to light up and mucking with how our LanguageServer understands file paths in Razor is a significantly large item. We will absolutely do it but given its size we wont have time to do it prior to locking down for \build.

:-/

I really want to believe this to be true.  But we've been on the receiving end of those same promises so many times in the past.  We can run this up our end.  But my requirement would be that we are allowed to set a date to remove this, and there is no ability to pushback on that work actually happening.  And problems caused by removal would be entirely on Razor and not on Roslyn.  i.e. if we can't RI because of this, it would be on Razor to drop everything to make it possible.

That said, i don't think it would be possible for people on our end to approve even that.  i.e. no one is going to want to approve something that potentially impacts our own ability to meet ship dates :-/

Reply:
> we are allowed to set a date to remove this, and there is no ability to pushback on that work actually happening

I'm absolutely with you here. Can we set May 31st as the absolute date, guaranteed no pushback from us?

> And problems caused by removal would be entirely on Razor and not on Roslyn. i.e. if we can't RI because of this, it would be on Razor to drop everything to make it possible.

Absolutely. In my original statement I even said, if this causes ANY other issues at all you can immediately remove it no questions asked.

> That said, i don't think it would be possible for people on our end to approve even that. i.e. no one is going to want to approve something that potentially impacts our own ability to meet ship dates :-/

If it'd make you feel better I'm willing to start that conversation with Jinu to get buy off from her as well. I see this as a super minor normalization that we intend to remove. Also, given that i'm the planner and executor on the Razor side I will make this happen.

Reply:
@jinujoseph and I talked this over and after May 31st if we've (ASP.NET Core Razor) have yet to fix the file path normalization issue on our end you're more than welcome to remove it. Also, if this file path normalization breaks any of the other Roslyn related experiences you are also welcome to remove the logic immediately. We understand the removal would break Razor's LSP diagnostic reporting and are OK with that happening. The product would be alpha quality and at most would be broken for a few weeks; more importantly we've already funded the reaction on our side and have confidence this will get done long before that date.

/cc @mkArtakMSFT